File: downloader.py
import json
import requests
import os
import codecs

URL = "https://www.marxists.org/"
def downloadFromJson(filename:str, outDir:str = 'www', dir:str = 'data'):
    with open(f'./{dir}/{filename}.json') as jsn:
        for i in json.load(jsn):
            if i['href'].endswith('.html') or i['href'].endswith('.htm'):
                r = requests.get(URL+i['href'])
                r.encoding = r.apparent_encoding
                if '<title>Object not found!</title>' not in r.text:
                    print(i['href'])
                    with codecs.open(os.path.join(outDir, i['href'].replace('/', '.')), 'wb') as f:
                        f.write(r.content)

#downloadFromJson('lenin')
#downloadFromJson('authors', 'authors')
downloadFromJson('glossary_index', 'glossary')

File: downloader.ts
import * as fs from 'fs/promises';
import axios from 'axios';
import * as path from 'path';
import { Index, Work, decode } from './common';

let files:string[] = [];


// Must check if is already installed
async function downloadFile(item:{href:string;}, outDir:string) {
    try {
        if ((item['href'].endsWith('.html') || item['href'].endsWith('.htm')) && !files.includes((/*URL +*/ item['href']).replace(/\//g, '.'))) {
            const response = await axios.get(/*URL + */item['href']);
            const htmlDecode = decode(response.data);

            if (!htmlDecode.includes('<title>Object not found!</title>')) {
                console.log(">", item['href']);
                const filePath = path.join(outDir,(/*URL + */item['href']).replace(/\//g, '.'));
                await fs.writeFile(filePath, htmlDecode);
            }
        }
    } catch (error) {
        console.error(`~${item['href']} ${error}`);
    }
}

async function downloadFromJson(filename: string, outDir: string = 'www', dir: string = 'data') {
    try {
        const jsonContents = await fs.readFile(`./${dir}/${filename}.json`, 'utf8');
        const data:(Work | Index)[] = JSON.parse(jsonContents);

        for(const item of data){ 
            if((item as Index).works)
                for(const work of (item as Index).works ?? [])
                    await downloadFile(work, outDir); 
            else            
                await downloadFile(item, outDir);
        }
    } catch (error) {
        console.error(`Error reading JSON file ${filename} : ${error}`);
    }
}

async function main() {
    files = await fs.readdir('./www3/');
    await downloadFromJson('lenin_works', 'www3');
    await downloadFromJson('stalin_works', 'www3');
    await downloadFromJson('mao_works', 'www3');
}

main();
